utils/time_slots_utils.js

```diff
@@ -0,0 +1,87 @@
+const {models} = require('../db/relations')
+const TimeSlot = models.TimeSlot
+
+const minDuration = 15
+const minToMsMult = 60000
+
+const TimeSlotsUtils = {
+
+    getIntersection: function (slots) {
+        let res = []
+        let pointers = new Array(slots.length).fill(0)
+        let l = -1
+        let r = -1
+        while (!isNaN(l) && !isNaN(r) ) {
+            let it_slots = new Array(slots.length).fill(0)
+            for (let i = 0; i < slots.length; i++) {
+                it_slots[i] = slots[i][pointers[i]]
+            }
+            let vals = it_slots.map(slot => slot !== undefined ? slot.startDatetime.getTime() : undefined)
+            l = Math.max(...vals)
+
+            vals = it_slots.map(slot => slot !== undefined ? slot.startDatetime.getTime() + slot.duration * minToMsMult : undefined)
+            r = Math.min(...vals)
+
+            if (l + minDuration <= r) {
+                res.push({startDatetime: new Date(l), duration: (new Date(r) - new Date(l)) / minToMsMult})
+            }
+
+            let valsCopy = [...vals]
+            while (valsCopy.indexOf(r) === valsCopy.lastIndex) {
+                valsCopy = valsCopy.filter(v => valsCopy.indexOf(v) !== r)
+                r = Math.min(...valsCopy)
+            }
+            pointers[vals.indexOf(r)]++
```

Comment:
Tutaj wydaje mi się, że jest coś pokićkane (albo nie ogarniam), bo np. są tu takie rzeczy jak **valsCopy.indexOf(v) !== r**, czyli porównywanie indexu z **r** które jest wartością z tablicy (bez sensu, przy okazji można zrobić **valsCopy.filter((v,i) => i !== r)**, indexOf kosztuje, a można go mieć z parametru, też lepiej nie używać w filter bezpośrednio kolekcji którą filtrujesz).

Też nie wiem jaki cel ma warunek w while'u, wartości w vals nie są w żaden sposób uporządkowane i nie wiem jaki związek ma index r'a z ostatnim indeksem w tablicy (tutaj też dobrze by było pominąć indexOf). 

Dla mnie się wydaje, że mogłoby to wyglądać tak 
```suggestion 
            vals.forEach((v, i) => {
                if(v < l + minDuration) pointer[i]++
            })
```

Reply:
Oczywiście nie mówię, że to na pewno dobre rozwiązanie, ale takie przesuwanie pointerów dla osób, których timeboxy kończą się wcześciej niż zaczyna się najpóźniejszy możliwy timebox wydaje się proste i sensowne.

Reply:
niestety wchodzi w infinite loop xd

Reply:
z tym indexOf racja - poprawiłem, ale niestety z proponowanym rozwiązaniem nie chce to działać. Jest to algorytm łatany trytytkami i pewnie dorobi się fixa, ale narazie na tyle na ile go otestowałem to działa xD

Reply:
> niestety wchodzi w infinite loop xd

Tego się obawiałem xD, ważne że działa w przypadkach testowych